### ISSUE TITLE
Hooks for retrying & failing without retry

### ISSUE THREAD

**Issue by thenovices** [lantins/resque-retry#123]
I don't think there are currently hooks implemented to run code in the case that (a) we failed but are retrying and (b) we failed and are not retrying. It would be nice to have these hooks.

For now, I've approximated this feature inside the `on_failure` hook with this guard:
```
unless respond_to?(:retry_criteria_valid?) && retry_criteria_valid?(e, *args)
   # ... code that runs if we are not retrying
end
```

Do you think it would be too hard to implement? As a first cut, we could do something less fancy than Resque, with its hook function prefixing (which is nice for plugins), and just call a class function if it exists, e.g., `on_try_again` and `on_give_up`

**Comment by jzaleski** [comment]
Hi @thenovices thank you for the suggestion. Did you review the `resque-retry` code to see where you might plug something like this in? I don't see a lot of harm in introducing this type of functionality if it can be done conditionally and cleanly. I'd be happy to help you work through this.

**Comment by thenovices** [comment]
Great to know that you would be interested. I was mainly looking at inserting the logic into `on_failure_retry`, around L422: https://github.com/lantins/resque-retry/blob/master/lib%2Fresque%2Fplugins%2Fretry.rb#L422

Basically I would fire `on_try_again_...` hooks when the condition is true, and `on_give_up_...` hooks when the condition is false.

Do those two hook prefixes sound good to you?

I'll send a PR in the next couple days; shouldn't be too difficult.

**Comment by lantins** [comment]
I don't quite understand the use-case for these hooks? or why using `on_failure` hooks along with some of the resque-retry helper methods like you explained is not sufficient for the job? Could you please enlighten me :)

**Comment by lantins** [comment]
@thenovices btw, good work on adding yardoc comments and a test case in your PR.

Reference the hook prefixes, I'm wondering if `on_will_retry` and `on_wont_retry` would be more clear?

**Comment by thenovices** [comment]
I personally found using the `retry_criteria_valid?` guard unintuitive, and I had to look through the source code to figure it out. I also wasn't 100% confident that that guard would catch all the cases. Having explicit hooks would make it much easier for the casual user to respond to those cases, I think.

I'm also totally fine with a different hook prefix -- I just picked whatever came to my mind first. My hesitations with `on_will_retry` and `on_wont_retry` is that (1) they're using future tense instead of present tense; and (2) they look really similar so it's just a tiny bit harder to differentiate at a quick glance.

Btw, if this all looks good, I'll update the README too.

**Comment by jzaleski** [comment]
There are a few things I would like to see changed. I will review this in depth as soon as I am able. Hopefully tonight.

**Comment by lantins** [comment]
@thenovices thanks for the feedback.

**Comment by thenovices** [comment]
Closed by #124
